activerecord/lib/active_record/associations/join_dependency/join_association.rb

```diff
@@ -63,11 +63,17 @@ def join_constraints(foreign_table, foreign_klass, join_type, alias_tracker)
                 end
               end
 
-              joins << join_type.new(table, Arel::Nodes::On.new(nodes))
-
-              if others && !others.empty?
-                joins.concat arel.join_sources
-                append_constraints(connection, joins.last, others)
+              if join_using_subquery?(scope)
+                arel.constraints.delete_at(0)
+                arel.constraints.unshift(*others)
+                joins << join_type.new(arel.as(table.name), Arel::Nodes::On.new(nodes))
```

Comment:
`#join_scope` returns an `ActiveRecord::Relation`. It would simplify things if we didn't need to infer which conditions are join constraints and which conditions are normal where values.

Here's an example:

```ruby
has_many :comments_with_subquery_where, -> { select(:post_id).where(id: 1) }
```

We might want this query to produce the following SQL:
```sql
INNER JOIN (SELECT comments.post_id FROM comments WHERE comments.id = 1) comments ON comments.post_id = posts.id
```

Instead, we get the following query, which fails with `no such column: comments.id`:
```sql
INNER JOIN (SELECT comments.post_id FROM comments) comments ON comments.id = 1 AND comments.post_id = posts.id
```

However, due to the way `others` works, this works fine:
```ruby
has_many :comments_with_subquery_where, -> { select(:post_id).where('comments.id = 1') }
```